src/libprojectM/MilkdropPresetFactory/SaltWaterTaffyPreset.cpp

```diff
@@ -0,0 +1,355 @@
+    {
+        start++;
+        end--;
+    }
+    if (end > start)
+        name = line.substr(start,end-start);
+    return PROJECTM_SUCCESS;
+}
+
+
+int parse_assignment_line(const std::string &line, std::string &name, std::string &value, std::string &comment)
+{
+    name.empty(); value.empty(); comment.empty();
+
+    size_t c = line.find("//");
+    if (c != std::string::npos)
+    {
+        comment = line.substr(c);
+    }
+    size_t eq = line.find('=');
+    if (eq == std::string::npos || eq > c)
+    {
+        name = line.substr(0,c);
+        trim(name);
+        if (name.empty())
+            return PROJECTM_SUCCESS;
+        if (PARSE_DEBUG) printf("parse taffy: expected line to containe '=': %s\n", line.c_str());
+        return PROJECTM_PARSE_ERROR;
+    }
+    name = line.substr(0,eq);
+    trim(name);
+    value = line.substr(eq+1);
+    trim(value);
+    return PROJECTM_SUCCESS;
+}
+
+
+
+// parse init conditions, looking for function blocks, until '}
+int parse_object(const char *prefix_init, const char *prefix_eqn, std::istream & fs, std::ostream & os)
+{
+    std::string line;
+    std::string name, value, comment;
+    char prefix_buffer[30];
+
+    while (getline(fs,line))
+    {
+        trim(line);
+        if (line.empty())
+        {
+            os << std::endl;
+            continue;
+        }
+        else if (line == "}")
+        {
+            os << std::endl;
+            return PROJECTM_SUCCESS;
+        }
+
+        int ret = parse_assignment_line(line, name, value, comment);
+        if (ret != PROJECTM_SUCCESS)
+            return ret;
+
+        if (value == "{")
+        {
+            if (name != "per_frame_init" && name != "per_frame" &&  name != "per_point" && name != "per_pixel")
+            {
+                if (PARSE_DEBUG) printf("parse taffy: block name should be one of frame_init, per_frame, per_point, per_frame, per_pixel: %s\n", line.c_str());
+                return PROJECTM_PARSE_ERROR;
+            }
+
+            if (!strlen(prefix_init) && (name == "per_frame" || name == "per_frame_init" || name == "per_pixel"))
+                sprintf(prefix_buffer, "%s%s_", prefix_eqn, name.c_str());
+            else
+                sprintf(prefix_buffer, "%s%s", prefix_eqn, name.c_str());
+
+            os << std::endl;
+            ret = parse_equations(prefix_buffer, fs, os);
+            if (ret != PROJECTM_SUCCESS)
+                return ret;
+        }
+        else if (value == "\"\"\"")
+        {
+            std::string shader="comp";
+            std::string lang="hlsl";
+            if (name != "comp_hlsl" && name != "comp_glsl_330" && name != "warp_hlsl" && name != "warp_glsl_330")
+            {
+                if (PARSE_DEBUG) printf("parse taffy: shader name should be one of comp_hlsl, comp_glsl_330, warp, or warp_glsl_330: %s\n", line.c_str());
+                return PROJECTM_PARSE_ERROR;
+            }
+            if (name == "comp_glsl_330" || name == "warp_glsl_330")
```

Comment:
the versions are highly likely to change. embedding these constants deep down in here is a very bad idea.